powaur.1.txt
.TH POWAUR 1 "10-06-2011" "powaur" "powaur manual"
.SH NAME
powaur \- an AUR helper with a pacman-like interface
.SH SYNOPSIS
.I powaur
<operation> [options] [targets]
.SH DESCRIPTION
powaur is an AUR helper which provides a pacman-like interface. Invoking
powaur involves specifying an operation with applicable options and targets.
.SH OPERATIONS
.TP
.B "-G, --getpkgbuild <package(s)>"
Download PKGBUILDS from the AUR. PKGBUILDS are downloaded as tarballs, then
extracted into the current working directory. Dependency resolution can be
turned on with the --deps flag.
.IP
powaur now supports multi-threaded downloading, up to a maximum of 10 threads.
.IP
NOTE: With --deps, downloading is threaded but dependency resolution is NOT
threaded.
.TP
.B "-Q, --query"
Query the pacman local database. This option lets you view information about
installed packages (eg. dependencies, provides, installed date). When no
packages are provided, all installed packages will be queried.
.TP
.B "-S, --sync"
Synchronize packages. When used with no options, at least one target is
required and powaur will attempt to install the target package(s) from the AUR.
Dependency resolution is now available.
.IP
Note that the use of this flag to install AUR packages is discouraged.
You are adviced to use the -G flag to download PKGBUILDS and review them
before installing them using makepkg.
.TP
.B "-B, --backup"
Backup the local pacman database. See Backup Usage.
.TP
.B "--crawl <package(s)>"
Shows a valid topological order for the given package(s). The package(s) can
be from local or sync databases or the AUR. For installed AUR packages, a fresh
copy of the PKGBUILD will be downloaded and dependency resolution is based on
the PKGBUILD instead of information in the local database.
.IP
NOTE: This option only shows a single valid topological order (there may be
many other valid topological orders) for the packages.
.TP
.B "-M, --maintainer <person>"
Search the AUR for packages maintained by person. 1 and only 1 maintainer must
be specified.
.TP
.B "--list-aur"
Lists all installed AUR packages.
.TP
.B "-h, --help"
Displays help message and exits.
.TP
.B "-V, --version"
Displays version information and exits.
.SH OPTIONS
.TP
.B "--color"
Enables colorized output. This option can be supplied many times but the effect
will be the same as if it was supplied once. If --nocolor is supplied, the
effect of this option is nullified.
.TP
.B "--debug"
Display debug messages.
.TP
.B "--nocolor"
Disables colorized output. This option can be supplied many times but the
effect will be the same as if it was supplied once. If --color is supplied, the
effect of this option is nullified.
.TP
.B "--noconfirm"
Bypass all questions. This option is passed down to makepkg.
.TP
.B "--threads <N>"
Limits powaur to spawn up to a maximum of N threads. Currently, mutli-threading
is limited to the -G operation. This option can be used to override the
"MaxThreads" setting in the configuration file.
.SH GETPKGBUILD OPTIONS
.TP
.B "--deps"
.br
Resolve dependencies for all target packages. If any dependencies are from the
AUR, they will be downloaded, with their dependencies resolved. This is done
until all dependencies have been resolved.
.IP
NOTE: For dependency resolution to work properly, the package names in the
PKGBUILDS must be exact. Use the --verbose flag to see more information for
dependency resolution.
.IP
NOTE: Dependency resolution is NOT threaded. But downloading of resolved packages
is threaded.
.TP
.B "--target <DIR>"
Downloads packages to alterante directory <DIR> instead of the current working
directory.
.SH QUERY OPTIONS
.TP
.B "-i, --info"
Displays information on installed packages. If no packages were specified, then
information on all installed packages is displayed.
.TP
.B "-s, --search"
Searches installed packages for the specified package(s). If no packages were
specified, then all installed packages will be listed.
.SH SYNC OPTIONS
.TP
.B "--check"
Only meaningful when used with -u. Will stop -Su before package upgrading. Use
this flag to check for outdated AUR packages without upgrading them.
.TP
.B "-i, --info"
Searches sync databases, followed by AUR for package information. If no
packages were specified, then information on all packages from all sync
databases is displayed.
.TP
.B "-s, --search"
Searches the AUR for a specified package. Currently, there is only support for
searching for 1 package.
.TP
.B "-u, --upgrade"
Checks all locally installed AUR packages and updates any outdated AUR packages,
with all relevant dependencies resolved.
.TP
.B "--vote"
.br
Used with -s. Orders search results from the AUR by vote count instead of
alphabetical order.
.SH MAINTAINER OPTIONS
.TP
.B "--vote"
.br
Orders search results from the AUR by vote count instead of alphabetical order.
.SH BACKUP USAGE
.IP "powaur -B"
Backup pacman database to current working directory.
.IP "powaur -B dir"
Backup pacman database to dir.
.SH Configuration
powaur looks for its configuration file first in:
.P
	$XDG_CONFIG_HOME/powaur.conf
.P
followed by:
.P
	$HOME/.config/powaur.conf
.P
1 and only 1 of the above files will be parsed. In the event of unspecified
configuration settings, powaur will fallback to using the defaults.
.P
A sample config file can be found at /usr/share/powaur/powaur.conf
.SH Colorized Output
By default, powaur's output is colorized. Thus, "color" starts with a value
of 1.
.P
The --color flag and "Color" option when set to "On" in powaur.conf increases
the value of "color" by 1.
.P
The --nocolor flag and "Color" option when set to "Off" in powaur.conf increases
the value of "color" by 1.
.P
If the final value of "color" is greater than 0, then output is colorized.
.SH CREDITS
Dave Reisner - For his kind permission to let powaur use code from cower
.P
pacman, git - Source code was used from both of these projects, especially
pacman
.SH BUG REPORTS
Please email bug reports to <pangyanhan@gmail.com>
.SH AUTHOR
Pang Yan Han <pangyanhan@gmail.com>


NEWS
VERSION         DESCRIPTION
--------------------------------------------------------------------------------
0.1.5         - --noconfirm option
              - --list-aur option to list installed AUR packages
              - -S: Only resolve deps for new / out-of-date AUR packages
0.1.4.1       - Allow installing of packages as root
              - Per user directories for downloading packages. Please comment out
			    the "TmpDir" option in your powaur.conf for this to take effect
			  - For -S, print immediate dependencies
			  - For -S, only print dependency graph when --debug is supplied
0.1.4         - -Su added to check for outdated AUR packages
              - --crawl operation to give valid topological order for packages
			  - Dependency resolution for -S
			  - Provides resolution for -G --deps
			  - --check option for -Su
0.1.3.1       - Handle '<' in PKGBUILDS
0.1.3         - Colorized output is now available!
0.1.2.1       - Update helptext to include --threads
0.1.2         - Multithreaded downloading of packages is now supported
              - Add "MaxThreads" option in config file
              - Add --threads option to specify maximum no. of threads.
              - Add --deps flag for -G. By default, dependency resolution is
			    turned off
0.1.1         - -G flag now resolves dependencies
              - Removed unsuccessfully downloaded packages
			  - Add verbose flag
0.1           - Initial release


powaur.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#include <getopt.h>

#include <alpm.h>
#include <curl/curl.h>
#include <yajl/yajl_parse.h>

#include "curl.h"
#include "download.h"
#include "environment.h"
#include "handle.h"
#include "json.h"
#include "package.h"
#include "powaur.h"
#include "sync.h"
#include "util.h"

static alpm_list_t *powaur_targets = NULL;

static int powaur_cleanup(int ret)
{
	FREELIST(powaur_targets);
	curl_cleanup();
	_pwhandle_free(pwhandle);
	cleanup_environment();
	alpm_release();

	exit(ret);
}

static int powaur_init(void)
{
	struct stat st;
	int ret = 0;

	if (alpm_initialize() == -1) {
		return -1;
	}

	if (setup_environment()) {
		return error(PW_ERR_INIT_ENV);
	}

	pwhandle = _pwhandle_init();
	if (!pwhandle) {
		return error(PW_ERR_INIT_HANDLE);
	}

	if (stat(powaur_dir, &st) != 0) {
		if (mkdir(powaur_dir, 0755)) {
			return error(PW_ERR_INIT_DIR);
		}
	}

	if (curl_init()) {
		return error(PW_ERR_CURL_INIT);
	}

	return ret;
}

static void usage(unsigned short op)
{
	if (op == PW_OP_MAIN) {
		printf("%s %s <operation> [...]\n", USAGE, MYNAME);
		printf("%s%s {-h --help}\n", TAB, MYNAME);
		printf("%s%s {-G --getpkgbuild} <%s>\n", TAB, MYNAME, PKG);
		printf("%s%s {-S --sync}        [%s] [%s]\n", TAB, MYNAME, OPT, PKG);
		printf("%s%s {-Q --query}       [%s] [%s]\n", TAB, MYNAME, OPT, PKG);
		printf("%s%s {-M --maintainer}  <%s>\n", TAB, MYNAME, PKG);
		printf("%s%s {-B --backup} [dir]\n", TAB, MYNAME);
		printf("%s%s {-V --version}\n", TAB, MYNAME);
		printf("%s%s --crawl <%s>\n", TAB, MYNAME, PKG);
		printf("%s%s --list-aur\n", TAB, MYNAME);
	} else {
		if (op == PW_OP_SYNC) {
			printf("%s %s {-S --sync} [%s] [%s]\n", USAGE, MYNAME, OPT, PKG);
		} else if (op == PW_OP_QUERY) {
			printf("%s %s {-Q --query} [%s] [%s]\n", USAGE, MYNAME, OPT, PKG);
		} else if (op == PW_OP_GET) {
			printf("%s %s {-G --getpkgbuild <%s>\n", USAGE, MYNAME, PKG);
		} else if (op == PW_OP_MAINTAINER) {
			printf("%s %s {-M --maintainer <%s>\n", USAGE, MYNAME, PKG);
		} else if (op == PW_OP_BACKUP) {
			printf("%s %s {-B --backup} [dir]\n", USAGE, MYNAME);
		} else if (op == PW_OP_LISTAUR) {
			printf("%s %s --list-aur\n", USAGE, MYNAME);
		}

		printf("%s:\n", OPT);

		switch (op) {
		case PW_OP_SYNC:
		case PW_OP_QUERY:
			printf("  -i, --info                 view package information\n");
			printf("  -s, --search <%s>  search %s repositories for %s\n",
				   PKG, op == PW_OP_SYNC ? "sync" : "local", PKG);
			break;
		case PW_OP_GET:
			printf("      --target <DIR>         downloads to alternate directory DIR\n");
			printf("      --deps                 resolves dependencies\n");
			printf("      --threads <N>          limit max no. of threads to N\n");
			break;
		case PW_OP_MAINTAINER:
			printf("      --vote                 order search results by votes\n");
			break;
		case PW_OP_CRAWL:
			printf("      --crawl <%s>    outputs dependency graph for %s\n", PKG, PKG);
			break;
		default:
			break;
		}

		if (op == PW_OP_SYNC) {
			printf("      --check                Works with -u, checks for outdated packages without upgrading\n");
			printf("  -u, --upgrade              Updates outdated AUR packages\n");
			printf("      --vote                 order search results by votes\n");
		}

		printf("      --debug                display debug messages\n");
		printf("      --verbose              display more messages\n");
		printf("      --color                Switches on color\n");
		printf("      --nocolor              Switches off color\n");
		printf("      --noconfirm            do not ask for any confirmation\n");
	}

cleanup:
	powaur_cleanup(1);
}

static void version()
{
	printf("powaur %s\n", POWAUR_VERSION);
	powaur_cleanup(0);
}

/* From pacman */
static int parsearg_op(int option, int dry_run)
{
	switch (option) {
	case 'G':
		if (dry_run) break;
		config->op = (config->op == PW_OP_MAIN ? PW_OP_GET : PW_OP_INVAL);
		break;
	case 'S':
		if (dry_run) break;
		config->op = (config->op == PW_OP_MAIN ? PW_OP_SYNC : PW_OP_INVAL);
		break;
	case 'Q':
		if (dry_run) break;
		config->op = (config->op == PW_OP_MAIN ? PW_OP_QUERY : PW_OP_INVAL);
		break;
	case 'M':
		if (dry_run) break;
		config->op = (config->op == PW_OP_MAIN ? PW_OP_MAINTAINER : PW_OP_INVAL);
		break;
	case 'B':
		if (dry_run) break;
		config->op = (config->op == PW_OP_MAIN ? PW_OP_BACKUP : PW_OP_INVAL);
		break;
	case PW_OP_CRAWL:
		if (dry_run) break;
		config->op = (config->op == PW_OP_MAIN ? PW_OP_CRAWL : PW_OP_INVAL);
		break;
	case 'h':
		if (dry_run) break;
		config->help = 1;
		break;
	case 'V':
		if (dry_run) break;
		config->version = 1;
		break;
	case PW_OP_LISTAUR:
		if (dry_run) break;
		config->op = (config->op == PW_OP_MAIN ? PW_OP_LISTAUR : PW_OP_INVAL);
		break;
	default:
		return -1;
	}

	return 0;
}

/* Parse options for -S (--sync) */
static int parsearg_sync(int option)
{
	switch (option) {
	case 'i':
		config->op_s_info = 1;
		break;
	case 's':
		config->op_s_search = 1;
		break;
	case 'u':
		config->op_s_upgrade = 1;
		break;
	case OPT_CHECK_ONLY:
		config->op_s_check = 1;
		break;
	default:
		return -1;
	}

	return 0;
}

/* Parse options for -Q (--query) */
static int parsearg_query(int option)
{
	switch (option) {
	case 'i':
		config->op_q_info = 1;
		break;
	case 's':
		config->op_q_search = 1;
		break;
	default:
		return -1;
	}

	return 0;
}

/* Parse options for -G (--getpkgbuild) */
static int parsearg_get(int option)
{
	switch (option) {
	case OPT_RESOLVE_DEPS:
		config->op_g_resolve = 1;
		break;
	default:
		return -1;
	}

	return 0;
}

/* Parse global arguments */
static int parsearg_global(int option)
{
	switch (option) {
	case OPT_DEBUG:
		config->loglvl |= PW_LOG_DEBUG;
		config->loglvl |= PW_LOG_VDEBUG;
		break;
	case OPT_SORT_VOTE:
		config->sort_votes = 1;
		break;
	case OPT_VERBOSE:
		config->verbose = 1;
		break;
	case OPT_TARGET_DIR:
		config->target_dir = strdup(optarg);
		break;
	case OPT_MAXTHREADS:
		config->opt_maxthreads = 1;
		powaur_maxthreads = atoi(optarg);
		break;
	case OPT_COLOR:
		if (!config->color_set) {
			++config->color;
			config->color_set = 1;
		}
		break;
	case OPT_NOCOLOR:
		if (!config->nocolor_set) {
			--config->color;
			config->nocolor_set = 1;
		}
		break;
	case OPT_NOCONFIRM:
		config->noconfirm = 1;
		break;
	default:
		return -1;
	}

	return 0;
}


/* Borrowed from pacman */
static int parseargs(int argc, char *argv[])
{
	int opt, option_index = 0;
	int res;

	const char *optstring = "BGMQSVhisuw";

	static struct option opts[] = {
		{"backup", no_argument, NULL, 'B'},
		{"getpkgbuild", no_argument, NULL, 'G'},
		{"maintainer", no_argument, NULL, 'M'},
		{"query", no_argument, NULL, 'Q'},
		{"sync", no_argument, NULL, 'S'},
		{"version", no_argument, NULL, 'V'},
		{"help", no_argument, NULL, 'h'},
		{"info", no_argument, NULL, 'i'},
		{"search", no_argument, NULL, 's'},
		{"upgrade", no_argument, NULL, 'u'},
		{"list-aur", no_argument, NULL, PW_OP_LISTAUR},
		{"check", no_argument, NULL, OPT_CHECK_ONLY},
		{"color", no_argument, NULL, OPT_COLOR},
		{"crawl", no_argument, NULL, PW_OP_CRAWL},
		{"debug", no_argument, NULL, OPT_DEBUG},
		{"nocolor", no_argument, NULL, OPT_NOCOLOR},
		{"noconfirm", no_argument, NULL, OPT_NOCONFIRM},
		{"vote", no_argument, NULL, OPT_SORT_VOTE},
		{"verbose", no_argument, NULL, OPT_VERBOSE},
		{"target", required_argument, NULL, OPT_TARGET_DIR},
		{"deps", no_argument, NULL, OPT_RESOLVE_DEPS},
		{"threads", required_argument, NULL, OPT_MAXTHREADS},
		{0, 0, 0, 0}
	};

	/* Get operation */
	while ((opt = getopt_long(argc, argv, optstring, opts, &option_index))) {
		if (opt < 0) {
			break;
		} else if (opt == 0) {
			continue;
		} else if (opt == '?') {
			return error(PW_ERR_OP_UNKNOWN);
		}

		parsearg_op(opt, 0);
	}

	if (config->op == PW_OP_INVAL) {
		return error(PW_ERR_OP_MULTI);
	} else if (config->version) {
		version();
	} else if (config->help) {
		usage(config->op);
	} else if (config->op == PW_OP_MAIN) {
		return error(PW_ERR_OP_NULL);
	}

	/* Parse remaining arguments */
	optind = 1;
	while ((opt = getopt_long(argc, argv, optstring, opts, &option_index))) {
		if (opt < 0) {
			break;
		} else if (opt == 0) {
			break;
			continue;
		} else if (opt == '?') {
			return error(PW_ERR_OP_UNKNOWN);
		} else if (parsearg_op(opt, 1) == 0) {
			/* Operation */
			continue;
		}

		switch (config->op) {
		case PW_OP_SYNC:
			res = parsearg_sync(opt);
			break;
		case PW_OP_QUERY:
			res = parsearg_query(opt);
			break;
		case PW_OP_GET:
			res = parsearg_get(opt);
			break;
		case PW_OP_MAINTAINER:
		case PW_OP_BACKUP:
		default:
			res = 1;
			break;
		}

		if (res == 0) {
			continue;
		}

		/* Parse global options */
		res = parsearg_global(opt);
		if (res != 0) {
			return error(PW_ERR_OP_UNKNOWN);
		}
	}

	/* GNU getopt permutes arguments. So leftover args are non-options
	 * and are taken to be packages
	 */
	while (optind < argc) {
		if (!alpm_list_find_str(powaur_targets, argv[optind])) {
			pw_printf(PW_LOG_DEBUG, "adding target: %s\n", argv[optind]);
			powaur_targets = alpm_list_add(powaur_targets, strdup(argv[optind]));
		}

		++optind;
	}

	return 0;
}

int main(int argc, char *argv[])
{
	int ret;

	if (setup_config()) {
		goto cleanup;
	}

	/* Parse arguments first so debug info can be up asap if enabled */
	ret = parseargs(argc, argv);
	if (ret) {
		goto cleanup;
	}

	/* If stdout is not terminal, turn off colourized output */
	if (!isatty(1)) {
		config->color = 0;
	}

	ret = powaur_init();
	ASSERT(ret == 0, goto cleanup);

	switch (config->op) {
	case PW_OP_GET:
		ret = powaur_get(powaur_targets);
		break;
	case PW_OP_SYNC:
		ret = powaur_sync(powaur_targets);
		break;
	case PW_OP_QUERY:
		ret = powaur_query(powaur_targets);
		break;
	case PW_OP_MAINTAINER:
		ret = powaur_maint(powaur_targets);
		break;
	case PW_OP_BACKUP:
		ret = powaur_backup(powaur_targets);
		break;
	case PW_OP_CRAWL:
		ret = powaur_crawl(powaur_targets);
		break;
	case PW_OP_LISTAUR:
		ret = powaur_list_aur();
		break;
	default:
		break;
	}

cleanup:
	powaur_cleanup(ret ? 1 : 0);
}
